openfl-tutorials/interactive_api/PyTorch_MVTec_PatchSVDD/workspace/PatchSVDD_with_Director.ipynb

```diff
@@ -0,0 +1,1170 @@
+   "execution_count": null,
+   "id": "67ae50de",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "federation.target_shape"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "920216d3",
+   "metadata": {
+    "scrolled": true
+   },
+   "outputs": [],
+   "source": [
+    "# First, request a dummy_shard_desc that holds information about the federated dataset \n",
+    "dummy_shard_desc = federation.get_dummy_shard_descriptor(size=10)\n",
+    "sample, target = dummy_shard_desc[0]\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "obvious-tyler",
+   "metadata": {},
+   "source": [
+    "## Creating a FL experiment using Interactive API"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "rubber-address",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from openfl.interface.interactive_api.experiment import TaskInterface, DataInterface, ModelInterface, FLExperiment"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "sustainable-public",
+   "metadata": {},
+   "source": [
+    "### Register dataset"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "unlike-texas",
+   "metadata": {},
+   "source": [
+    "We extract User dataset class implementation.\n",
+    "Is it convinient?\n",
+    "What if the dataset is not a class?"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "dd2407cf",
+   "metadata": {
+    "scrolled": true
+   },
+   "outputs": [],
+   "source": [
+    "#List of objects in MVTec dataset\n",
+    "\n",
+    "objs = ['bottle', 'cable', 'capsule', 'carpet', 'grid', 'hazelnut',\n",
+    "        'leather', 'metal_nut', 'pill', 'screw', 'tile', 'toothbrush',\n",
+    "        'transistor', 'wood', 'zipper']\n",
+    "#Arguments\n",
+    "args = {\n",
+    "'obj' : 'bottle',\n",
+    "'lambda_value': '1e-3',\n",
+    "'D' : 64,\n",
+    "'lr' : '1e-4',\n",
+    "}\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "8669bc63",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import argparse\n",
+    "import torch\n",
+    "from functools import reduce\n",
+    "from torch.utils.data import DataLoader,Dataset\n",
+    "import torch.nn as nn\n",
+    "import torch.nn.functional as F\n",
+    "import math\n",
+    "import numpy as np\n",
+    "from PIL import Image\n",
+    "from imageio import imread\n",
+    "from glob import glob\n",
+    "from sklearn.metrics import roc_auc_score\n",
+    "import os, shutil\n",
+    "import _pickle as p\n",
+    "from contextlib import contextmanager\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "64f37dcf",
+   "metadata": {
+    "scrolled": true
+   },
+   "outputs": [],
+   "source": [
+    "import PIL\n",
+    "from torch.utils.data import Dataset, DataLoader, SubsetRandomSampler\n",
+    "from torchvision import transforms as tsf\n",
+    "from utils import *\n",
```

Comment:
Should avoid using `import *`. Use `import utils` and then `utils.to_device` to keep context or 'from utils import to_device`.

Reply:
done